### PR TITLE
Sti evolving entities

### PR DESCRIPTION
**Pull request by stephenh**

### CODE REVIEW

File: content/drafts/using-sti-to-evolve-entities.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+  // Also well as any "incoming" FKs that depend on task
+  billLineItems?: BillLineItem[];
+  homeownerContractDraws?: HomeownerContractDraw[];
+  requiredTaskDocuments?: RequiredTaskDocument[];
+  documents?: Document[];
+  jobLogNotes?: JobLogNote[];
+}
+```
+
+Joist also produces a "reactivity report" that further shows how `Task` is connected to the rest of the system, beyond the immediate foreign keys, via Joist's reactive rules & reactive fields features:
```
Comment (blambillotte) [inline content/drafts/using-sti-to-evolve-entities.md]: Question: is this something joist provides out of the box? Seems like we have our own script to read the config and print this `reactions` file?

Comment (stephenh) [reply content/drafts/using-sti-to-evolve-entities.md]: You caught me hand-waving on this :-) but yes, the reactions is still an internal-only script atm.

File: content/drafts/using-sti-to-evolve-entities.md [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+</div>
+
+This audit was great because it gave us context & concrete data points for then evaluating our high-level approaches, instead of just our initial gut feelings. 
+
+## High Level Approaches
+
+We generally had three approaches to consider:
+
+1. Create an entirely new table/entity, like `tasks_v2` / `TasksV2`, or
+2. Reuse the existing `tasks` / `Task` table/entity, and somehow manage the v1/v2 differentiation
```
Comment (blambillotte) [inline content/drafts/using-sti-to-evolve-entities.md:126]:
```suggestion
2. Reuse the existing `tasks` / `Task` table/entity, and somehow manage the v2/v3 differentiation
```
maybe not an important distinction, but you do point out above that this is v3?

Comment (blambillotte) [reply content/drafts/using-sti-to-evolve-entities.md:126]: Ehh I see there are more mentions of v2 below, maybe simpler to hope the reader doesn't care about that distinction

Comment (stephenh) [reply content/drafts/using-sti-to-evolve-entities.md:126]: Yep...I should probably just pretend that "the real v3" is "just v2" for the purposes of the blog post.

File: content/drafts/using-sti-to-evolve-entities.md [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+   * Pro: Doesn't break any existing data/Looker reports that use `tasks`
+   * Con: Requires making task v1/v2-specific columns nullable (shared table)
+   * Con: No code-level differentiation between v1 & v2 tasks
+
+2. Use [Class Table Inheritance](https://joist-orm.io/docs/advanced/class-table-inheritance) to model `TaskV1` and `TaskV2` as subtypes of `Task`, with a shared `tasks` "base table" and separate `tasks_v1` and `tasks_v2` sub-tables.
+
+   We like judiciously using CTI, because it has very clean modeling of "this entity has polymorphic subtypes" within both the database schema and the domain model. For example, it lets each subtype have strong database-level constraints and subtype-specific FKs (both outgoing & incoming); (the Joist STI documentation goes more into these details).
+   
+   However, needing to move the v1 columns into a `tasks_v1` table (to satisfy CTI's table-per-subtype) would defeat the goal of minimizing outside observability of the change (particularly to the data lake), and also we don't really want "TaskV1" and "TaskV2" concepts in our domain model, which is what CTI is best suited for. We're looking more of an implementation hack.
```
Comment (blambillotte) [inline content/drafts/using-sti-to-evolve-entities.md]: "We're looking more" - looking for? looking at?

File: content/drafts/using-sti-to-evolve-entities.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+* Evaluating whether create-new or evolve-existing domain entities
+  * Look at number of usages; if there's more than a handful, probably evolve 
+  * Consider whether this is really a new domain concept, or just an evolution of an existing one
+* Evaluating whether to use STI or CTI for "v1/v2" entity evolution
+  * CTI is more pure, but STI can be more pragmatic
+  * CTI is better for new domain concepts, STI is better for evolving implementations
+
+## Acknowledgments
+
+Many thanks to Ben Lamboitte and Zach Gavin, who were instrumental in researching, evaluating, and prototyping the "new or evolve?" approaches & trade-offs, and Ben in particular for doing basically all the "take theory to practice" work in rolling these changes out to the system.
```
Comment (blambillotte) [inline content/drafts/using-sti-to-evolve-entities.md]: 😆 
```suggestion
Many thanks to Ben Lambillotte and Zach Gavin, who were instrumental in researching, evaluating, and prototyping the "new or evolve?" approaches & trade-offs, and Ben in particular for doing basically all the "take theory to practice" work in rolling these changes out to the system.
```

Comment (stephenh) [reply content/drafts/using-sti-to-evolve-entities.md]: Lol! I didn't have slack auto-complete :see_no_evil: ty!

**Pull request merged**